### PR TITLE
kubelet/eviction: eliminate redundant allocations when handling eventfd

### PR DESCRIPTION
**Pull request by panjf2000**
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

Eliminate the unnecessary memory allocations when reading from eventfd.

**Which issue(s) this PR fixes**:

None

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
NONE
```

### CODE REVIEW

**Comment by panjf2000** [comment]
@mrunalp @krmayankk Have some time to take a look on this?

**Comment by panjf2000** [comment]
Please take a look at this PR, thanks! @sjenning @mrunalp @krmayankk

**Comment by panjf2000** [comment]
Could somebody take a glance at this? It's been a week!

**Comment by panjf2000** [comment]
The new pattern to create epoll instance is more robust, which is widely used among several world-renowned open-source projects: 
- [libuv](https://github.com/libuv/libuv/blob/30ff5bf2161257921f3a3ce5655804f7cb282aa9/src/unix/linux-core.c#L88-L98)
- [libevent](https://github.com/libevent/libevent/blob/852af060af7c7e439f0db97829ec74bedeba64d5/epoll.c#L179-L187)
- [go](https://github.com/golang/go/blob/0e85fd7561de869add933801c531bf25dee9561c/src/runtime/netpoll_epoll.go#L33-L41)

**Comment by panjf2000** [comment]
I almost forgot about this PR, it's been two months...
Why exactly hasn't anyone taken a look at it so far?
@mrunalp @krmayankk

**Comment by ehashman** [comment]
> I almost forgot about this PR, it's been two months...
> Why exactly hasn't anyone taken a look at it so far?

Your PR doesn't fix any reported issues in k8s and thus has backlog (lowest) priority, while also being technically complex to review. We have over 150 PRs on our backlog for SIG Node alone and a limited reviewer/approver pool.

I recommend you reach out in #pr-reviews on k8s slack if you want to get reviewer attention.

**Comment by nikhita** [comment]
cc @adisky

**Comment by sftim** [comment]
@panjf2000 thanks for your patience so far

File: pkg/kubelet/eviction/threshold_notifier_linux.go
Comment (fromanirh) [inline pkg/kubelet/eviction/threshold_notifier_linux.go:61]: looking at code and at manpages, this is the only change that makes me think, I'll need to dig into the history and into the codebase to learn why we didn't introduce this flag earlier.

**Review by ffromani — COMMENTED** [review]
I need to dig more to learn about the NONBLOCKING flag - most notably, why we haven't before. Everything else seems clean and nice.

File: pkg/kubelet/eviction/threshold_notifier_linux.go
Comment (fromanirh) [reply pkg/kubelet/eviction/threshold_notifier_linux.go:61]: Sure: if you could elaborate what's the benefit of this change - just this change, just adding this flag in the context of this code, that will make things faster. At glance (still digging), considering how the code behaves in `wait()` (https://github.com/kubernetes/kubernetes/blob/v1.20.4/pkg/kubelet/eviction/threshold_notifier_linux.go#L141) I don't see yet an obvious benefit.

File: pkg/kubelet/eviction/threshold_notifier_linux.go
Comment (fromanirh) [inline pkg/kubelet/eviction/threshold_notifier_linux.go:76]: how old though? Do you have an example of distribution (or at least kernel) so old that kubernetes supports?
Asking because epoll_create1 dates back to kernel 2.6.27 (!!) and glibc 2.9 (https://man7.org/linux/man-pages/man2/epoll_create1.2.html) , I don't think k8s can actually run on so ancient platforms.

File: pkg/kubelet/eviction/threshold_notifier_linux.go
Comment (SergeyKanzhelev) [reply pkg/kubelet/eviction/threshold_notifier_linux.go:61]: Any updates on this?

File: pkg/kubelet/eviction/threshold_notifier_linux.go
Comment (panjf2000) [reply pkg/kubelet/eviction/threshold_notifier_linux.go:76]: What's the minimum kernel version supported by k8s?

File: pkg/kubelet/eviction/threshold_notifier_linux.go
Comment (panjf2000) [reply pkg/kubelet/eviction/threshold_notifier_linux.go:61]: > Sure: if you could elaborate what's the benefit of this change - just this change, just adding this flag in the context of this code, that will make things faster. At glance (still digging), considering how the code behaves in `wait()` (https://github.com/kubernetes/kubernetes/blob/v1.20.4/pkg/kubelet/eviction/threshold_notifier_linux.go#L141) I don't see yet an obvious benefit.

It's more like a convention flag, it makes the `write` to the inside uint64_t counter of `eventfd` never block, of course, in theory.

File: pkg/kubelet/eviction/threshold_notifier_linux.go
Comment (fromanirh) [reply pkg/kubelet/eviction/threshold_notifier_linux.go:76]: The information is susprisingly scarce. Roughly 3 years ago the minimum documented requirement was kernel 3.10: https://github.com/kubernetes/kubernetes/issues/7528

Comment (panjf2000) [reply pkg/kubelet/eviction/threshold_notifier_linux.go:76]: Then this part of code changes can be remove from this PR.

File: pkg/kubelet/eviction/threshold_notifier_linux.go
Comment (fromanirh) [reply pkg/kubelet/eviction/threshold_notifier_linux.go:61]: So there is not a performance gain nor a bugfix rationale for this change? I'd remove from this set then.

File: pkg/kubelet/eviction/threshold_notifier_linux.go
Comment (fromanirh) [reply pkg/kubelet/eviction/threshold_notifier_linux.go:76]: ok, please remove this specific change please.

File: pkg/kubelet/eviction/threshold_notifier_linux.go
Comment (panjf2000) [reply pkg/kubelet/eviction/threshold_notifier_linux.go:61]: Actually, I think `unix.EFD_CLOEXEC` is also a convention flag here without actual impacts?

**Comment by SergeyKanzhelev** [comment]
this PR is now just a move of a buffer out of the loop. Please squash commits

**Comment by panjf2000** [comment]
> @panjf2000: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:
> 
> Test name	Commit	Details	Required	Rerun command
> pull-kubernetes-e2e-gce-ubuntu-containerd	[3033a64](https://github.com/kubernetes/kubernetes/commit/3033a641356dbb744caa8f683aff41b5202b309e)	[link](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/96828/pull-kubernetes-e2e-gce-ubuntu-containerd/1456165010089709568/)	true	`/test pull-kubernetes-e2e-gce-ubuntu-containerd`
> [Full PR test history](https://prow.k8s.io/pr-history?org=kubernetes&repo=kubernetes&pr=96828). [Your PR dashboard](https://prow.k8s.io/pr?query=is%3Apr%20state%3Aopen%20author%3Apanjf2000). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/kubernetes/kubernetes/issues?q=is:issue+is:open) when you hit one in your PR.

image

Any clues?

**Comment by dims** [comment]
@SergeyKanzhelev looks like there's a test failure :(

**Comment by panjf2000** [comment]
image

It didn't seem to be a relevant failure?

**Comment by panjf2000** [comment]
rerun the tests? @SergeyKanzhelev @fromanirh

**Comment by fromanirh** [comment]
the e2e lanes look good, not sure what's going on with EasyCLA

**Comment by panjf2000** [comment]
> the e2e lanes look good, not sure what's going on with EasyCLA

Anything I can do to fix it?

**Comment by fromanirh** [comment]
> > the e2e lanes look good, not sure what's going on with EasyCLA
> 
> Anything I can do to fix it?

We probably need to retrigger the lane. I don't have permissions to do so. Uploading again the change, perhaps after a rebase on top of current branch, may unblock it.

**Comment by SergeyKanzhelev** [comment]
let me try to close and re-open the PR. Sometimes it helps

/close

**Review by SergeyKanzhelev — APPROVED** [review]
it is a trivial change in it's current form

/lgtm

**Comment by panjf2000** [comment]
> @panjf2000: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:
> 
> Test name	Commit	Details	Required	Rerun command
> pull-kubernetes-e2e-kind	[3033a64](https://github.com/kubernetes/kubernetes/commit/3033a641356dbb744caa8f683aff41b5202b309e)	[link](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/96828/pull-kubernetes-e2e-kind/1496191801000726528)	true	`/test pull-kubernetes-e2e-kind`
> pull-kubernetes-integration	[3033a64](https://github.com/kubernetes/kubernetes/commit/3033a641356dbb744caa8f683aff41b5202b309e)	[link](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/96828/pull-kubernetes-integration/1496191800958783488)	true	`/test pull-kubernetes-integration`
> [Full PR test history](https://prow.k8s.io/pr-history?org=kubernetes&repo=kubernetes&pr=96828). [Your PR dashboard](https://prow.k8s.io/pr?query=is%3Apr%20state%3Aopen%20author%3Apanjf2000). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/kubernetes/kubernetes/issues?q=is:issue+is:open) when you hit one in your PR.

Looked like two unrelated failures?

**Comment by panjf2000** [comment]
Everything seems fine, could we merge it?
@fromanirh @SergeyKanzhelev

**Comment by fromanirh** [comment]
> Everything seems fine, could we merge it? @fromanirh @SergeyKanzhelev

You got a LGTM, but you need an approver for the change: see https://github.com/kubernetes/kubernetes/pull/96828#issuecomment-1048097706

**Comment by panjf2000** [comment]
> > Everything seems fine, could we merge it? @fromanirh @SergeyKanzhelev
> 
> You got a LGTM, but you need an approver for the change: see [#96828 (comment)](https://github.com/kubernetes/kubernetes/pull/96828#issuecomment-1048097706)

@mrunalp @sjenning

**Comment by panjf2000** [comment]
@Random-Liu @dchen1107 @derekwaynecarr @yujuhong @klueska 
Please take a look at this whoever is available, thanks~

**Review by mrunalp — APPROVED** [review]

**Pull request merged**
